### PR TITLE
I/O repos

### LINKED ISSUE #23: Thoughts about the input/ouput folders structure and design experiment
I think we should separate the input and output folders which can share the same root directory, e.g. `root_dir = '../data/streetscene'`.

* input folder with the raw datasets
  - `input/mapillary` where you can find the `tranining` and `validation` folders and the further preprocess folders (resized and/or aggregated images)
  - `input/shapes`

* output folder with the different output result for the different dataset
  - `output/mapillary`
  - `output/shapes`

and create a folder for each run. A number of parameter names and the date can occur in the folder name. A meta.json or YAML file inside this folder which summarizes the list of input parameters and hyptohesis.

### PR DESCRIPTION
**Pull request by delhomer**
Modify the structure of input/output repositories. The new architecture is described in issue #23 .

The goal of this PR is to distinguished each instance (instance name, image size, network size, label aggregation, dropout rate, learning rate) when creating model backups, as well as to distinguish clearly between research problems (for instance feature detection is the only problem addressed in `master`, however semantic segmentation will be merged soon, see #19 ). The PR sorts as clearly as possible preprocessed data too, in a dedicated directory : a preprocessed data set is essentially characterized by image size and label aggregation.

Fix #23.

### CODE REVIEW

File: sources/test.py [hunk 1]
```diff
@@ -73,15 +79,27 @@
...
+
     # Data path and repository management
     dataset_repo = os.path.join(args.datapath, args.dataset)
-    testing_name = "testing_" + str(image_size)
-    os.makedirs(os.path.join(dataset_repo, testing_name, "images"), exist_ok=True)
-    testing_filename = os.path.join(dataset_repo, testing_name + '.json')
+    input_repo = os.path.join(dataset_repo, "input")
+    preprocessed_repo = str(image_size) + "_" + aggregate_value
+    preprocessed_path = os.path.join(dataset_repo, "preprocessed", preprocessed_repo)
+    testing_filename = os.path.join(preprocessed_path, "testing.json")
```
Comment (garaud) [inline sources/test.py]: What is the purpose of the `testing.json` file?

File: sources/train.py [hunk 2]
```diff
@@ -150,27 +154,39 @@
...
-    suffix = "" if not args.aggregate_label else "aggregate_"
-    training_name = "training_" + suffix + str(args.image_size)
-    validation_name = "validation_" + suffix + str(args.image_size)
-    os.makedirs(os.path.join(dataset_repo, training_name, 'images'), exist_ok=True)
-    os.makedirs(os.path.join(dataset_repo, training_name, 'labels'), exist_ok=True)
-    os.makedirs(os.path.join(dataset_repo, validation_name, 'images'), exist_ok=True)
-    os.makedirs(os.path.join(dataset_repo, validation_name, 'labels'), exist_ok=True)
-    training_filename = os.path.join(dataset_repo, training_name + '.json')
-    validation_filename = os.path.join(dataset_repo, validation_name + '.json')
+    input_repo = os.path.join(dataset_repo, "input")
```
Comment (garaud) [inline sources/train.py]: Maybe you can put this content into a dedicated function e.g. `prepare_folders` or something else. What do you think?

File: sources/utils.py [hunk 3]
```diff
@@ -377,3 +377,18 @@ def split_list(l, nb_partitions):
     """
     return [l[i:(i+len(l)/nb_partitions)]
             for i in range(0, len(l), int(len(l)/nb_partitions))]
+
+def list_to_str(l):
+    """Transform the input list into a ready-to-print string
+
+    Parameters
+    ----------
+    l: list
```
Comment (garaud) [inline sources/utils.py]: Add a whitespace between the argument and its type.

File: sources/utils.py [hunk 4]
```diff
@@ -377,3 +377,18 @@ def split_list(l, nb_partitions):
     """
     return [l[i:(i+len(l)/nb_partitions)]
             for i in range(0, len(l), int(len(l)/nb_partitions))]
+
+def list_to_str(l):
```
Comment (garaud) [inline sources/utils.py]: I prefer having a variable name with more than one letter here. I like `seq` for *sequence*. It can be a list, a tuple, etc. It's a collection of strings. Or another name > 1 letter :)

File: sources/utils.py [hunk 5]
```diff
@@ -377,3 +377,18 @@ def split_list(l, nb_partitions):
...
+    ----------
+    l: list
+        Input list that must be transformed
+
+    Returns
+    -------
+    str
+        Printable version of input list
+    """
+    return '-'.join([str(i) for i in l])
```
Comment (garaud) [inline sources/utils.py]: You can also avoid the comprehension list and write

```python
'-'.join(str(i) for i in seq)
```

File: sources/utils.py (hunk 4 above)
Comment (garaud) [inline sources/utils.py]: It's really not necessary. You can imagine to have a more generic function if you pass a "separator" parameter, with the value `-` by default.

```python
def list_to_str(seq, sep='-'):
    pass
```

File: sources/train.py [hunk 6]
```diff
@@ -150,27 +154,39 @@
...
+    preprocessed_validation_path = os.path.join(preprocessed_path, "validation")
+    backup_path = os.path.join(dataset_repo, "output", args.model)
+    os.makedirs(os.path.join(preprocessed_training_path, "images"), exist_ok=True)
+    os.makedirs(os.path.join(preprocessed_training_path, "labels"), exist_ok=True)
+    os.makedirs(os.path.join(preprocessed_validation_path, "images"), exist_ok=True)
+    os.makedirs(os.path.join(preprocessed_validation_path, "labels"), exist_ok=True)
+    os.makedirs(backup_path, exist_ok=True)
+
+    # Instance name (name + image size + network size + batch_size + aggregate? + dropout + learning_rate)
+    instance_name = (args.name + "_" + str(args.image_size) + "_" + args.network_size
```
Comment (garaud) [inline sources/train.py]: Maybe you can put all these variables in a list and call the `list_to_str` function with a `_` separator

File: sources/test.py [hunk 7]
```diff
@@ -34,6 +34,8 @@
 if __name__ == '__main__':
     parser = argparse.ArgumentParser(description=("Convolutional Neural Netw"
                                                   "ork on street-scene images"))
+    parser.add_argument('-a', '--aggregate-label', action='store_true',
```
Comment (garaud) [inline sources/test.py:37]: I think, for further pull requests, we can factorize some code between the `train.py` and `test.py` modules because I see multiple duplication: the parsing of the command line, folders and files checking, etc.

File: sources/dataset.py [hunk 8]
```diff
@@ -186,11 +186,11 @@ def group_image_label(self, image):
                 a[mask] = root_id
         return Image.fromarray(a, mode=image.mode)
 
-    def _preprocess(self, datadir, image_filename, aggregate, labelling=True):
+    def _preprocess(self, image_filename, output_dir, aggregate, labelling=True):
         """Resize/crop then save the training & label images
 
-        :param datadir: str
         :param image_filaname: str
+        :param output_dir: str
```
Comment (garaud) [inline sources/dataset.py:193]: Is is possible to update the docstring format? With numpydoc style?

**Review by garaud — CHANGES_REQUESTED** [review]

File: sources/test.py (hunk 1 above)
Comment (delhomer) [reply sources/test.py]: The same as `training.json` and `validation.json`: initialize the `Dataset` object that contains all the dataset image characteristics. Even if it is the testing phase, we still have a directory full of thousands of images. And we need to centralize the information so as to read them carefully.

The "labels" key within this file appears as unnecessary, however we need to store dummy labels because of TensorFlow feeding process (labels are expected during image batching, see `define_batch` method in `model.py`). This behavior could certainly be improved, however it seems more accurate to me to open a new dedicated issue.

File: sources/utils.py (hunk 3 above)
Comment (delhomer) [reply sources/utils.py]: Done.

File: sources/utils.py (hunk 4 above)
Comment (delhomer) [reply sources/utils.py]: Done.

File: sources/utils.py (hunk 5 above)
Comment (delhomer) [reply sources/utils.py]: Done.

File: sources/utils.py (hunk 4 above)
Comment (delhomer) [reply sources/utils.py]: The `sep` argument has been added, however I'm not sure to understand your point, when you write "it's really not necessary". Do you mean that this function is useless?

File: sources/train.py (hunk 6 above)
Comment (delhomer) [reply sources/train.py]: Could be possible. Not sure it is a fundamental improvement, however it would be cleaner, for sure.

File: sources/test.py (hunk 7 above)
Comment (delhomer) [reply sources/test.py:37]: Be careful, the command line parsing is not managed exactly in the same manner in both files. The point is that we must point to the accurate `<datadir>/<dataset>/output/<problem>/checkpoints/<instance_name>` checkpoint to recover the model. Either we set all arguments as in `train.py`, or we set only one argument, that is the whole `<instance_name>` string. I chose this second way, as some arguments are not useful in `test.py`. Do you think the first way is better?

File: sources/dataset.py (hunk 8 above)
Comment (delhomer) [reply sources/dataset.py:193]: I suggest to open another issue to address this.

File: sources/train.py (hunk 2 above)
Comment (delhomer) [reply sources/train.py]: Why not. It will clarify the `train.py` and `test.py` modules. These tedious code lines seems unavoidable though (let's hide them into `utils.py` :) ).

File: sources/utils.py (hunk 4 above)
Comment (garaud) [reply sources/utils.py]: It's not necessary "for now", but I think it's a good design. See a comment above with the list of args and the `_` separator. We already have an good usage for this function with the `sep` parameter!

File: sources/train.py (hunk 6 above)
Comment (garaud) [reply sources/train.py]: Agreed.

File: sources/dataset.py (hunk 8 above)
Comment (garaud) [reply sources/dataset.py:193]: ok

File: sources/test.py (hunk 7 above)
Comment (garaud) [reply sources/test.py:37]: I'm aware that all arguments parsing can not be shared between `train` and `test`. It's just a feeling that some stuff could be factorized when I saw the current diffs. I wonder if it's relevant.

**Review by garaud — APPROVED** [review]
Sounds good to me. You can merge.

File: sources/test.py (hunk 7 above)
Comment (delhomer) [reply sources/test.py:37]: You're right, it is relevant. The dataset creation and populating, as well as the glossary management, are common. The only parts that must be differentiated are argument parsing and model calling. Do we address it in this PR?

**Comment by garaud** [comment]
No. We can merge as it is.
Raphael Delhome writes:

> delhomer commented on this pull request.
>
>
>
>> @@ -34,6 +34,8 @@
>  if __name__ == '__main__':
>      parser = argparse.ArgumentParser(description=("Convolutional Neural Netw"
>                                                    "ork on street-scene images"))
> +    parser.add_argument('-a', '--aggregate-label', action='store_true',
>
> You're right, it is relevant. The dataset creation and populating, as well as the glossary management, are common. The only parts that must be differentiated are argument parsing and model calling. Do we address it in this PR?


-- 
Damien Garaud
Data Scientist – Developer, Oslandia
http://oslandia.com
